[Feature][API] New restful API for workflow and schedule

zhongjiajie opened the pull request:
CURD for workflow and schedule, different with exists
API, this new restful api only operate single resource
in each request, and return the latest. For example,
previous workflow should also need to post tasks definition
and tasks relation definition, but this patch will allow
you to create workflow without task relate information

zhongjiajie commented:
This PR is not equal to refactored exists API as #10257 do, we also separate existing`process definition` and `schedule` into single resources CRUD(previous process definition also will submit tasks and task relation). And we make it more Restful

zhongjiajie committed:
[feat] New restful API for workflow and schedule

CURD for workflow and schedule, different with exists
API, this new restful api only operate single resource
in each request, and return the latest. For example,
previous workflow should also need to post tasks definition
and tasks relation definition, but this patch will allow
you to create workflow without task relate information

zhongjiajie committed:
Add license

zhongjiajie committed:
fix code scaning

zhongjiajie committed:
use checkProjectAndAuthThrowException, and fix CI error

zhongjiajie committed:
smell code

zhongjiajie committed:
recover code scanning fix
fix the warning code

zhongjiajie committed:
fix conflict

zhongjiajie commented:
fix conflict from adding log in #11782

zhongjiajie committed:
remove duplicate logger

zhongjiajie commented:
PTAL if you have time @caishunfeng @SbloodyS

caishunfeng commented on dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/ScheduleV2Controller.java:
```diff
@@ -0,0 +1,169 @@
+    @AccessLogAnnotation(ignoreRequestArgs = "loginUser")
+    public ResourceResponse getSchedules(@ApiIgnore @RequestAttribute(value = Constants.SESSION_USER) User loginUser,
+                                         @PathVariable("id") Integer id) {
+        Schedule schedule = schedulerService.getSchedules(loginUser, id);
```
```suggestion
        Schedule schedule = schedulerService.getSchedule(loginUser, id);
```

caishunfeng replied:
Or change into multi method.

caishunfeng commented on dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/dto/schedule/ScheduleCreateRequest.java:
```diff
@@ -0,0 +1,120 @@
+    @ApiModelProperty(example = "environment-code")
+    private long environmentCode;
+
+    public String getScheduleParam() {
```
Why not use json util?

ruanwenjun replied:
It's better to create a pojo named ScheduleParam with this 4 fields.

zhongjiajie replied:
> Why not use json util?
can you give some example about how to use it?

zhongjiajie replied:
>It's better to create a pojo named ScheduleParam with this 4 fields.
@ruanwenjun  I want keep parameter same leve, as it in table `t_ds_scheldule`. If we add pojo, we have to accept one parameter like 

```json
{
    "releaseState": "Online",
    "ScheduleParam": "{'startTime': startTime, 'endTime': endTime, 'crontab': contab}"
}
```

but we I want it is accept parame like 

```json
{
    "releaseState": "Online",
    "startTime": "startTime",
    "endTime": "endTime",
    "crontab": "contab"
}
```

zhongjiajie replied:
Oh I get your point, we can create pojo and then 

```java
ScheduleParam scheduleParam = new ScheduleParam(this.startTime, this.endTime, this.crontab, this.timezoneId);
return scheduleParam.toString();
```

right?

zhongjiajie replied:
I currently use 

```java
Gson gson = new GsonBuilder().serializeNulls().create();
ScheduleParam scheduleParam = new ScheduleParam(this.startTime, this.endTime, this.crontab, this.timezoneId);
return gson.toJson(scheduleParam);
```

to covert pojo into json string

caishunfeng replied:
> I currently use
> 
> ```java
> Gson gson = new GsonBuilder().serializeNulls().create();
> ScheduleParam scheduleParam = new ScheduleParam(this.startTime, this.endTime, this.crontab, this.timezoneId);
> return gson.toJson(scheduleParam);
> ```
> 
> to covert pojo into json string

I think it's OK, cc @ruanwenjun

caishunfeng commented on dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/dto/schedule/ScheduleUpdateRequest.java:
```diff
@@ -0,0 +1,133 @@
+    }
+
+    public Schedule mergeIntoSchedule(Schedule schedule) {
+        Schedule scheduleDeepCopy = JSONUtils.parseObject(JSONUtils.toJsonString(schedule), Schedule.class);
```
use BeanUtils.cloneBean

zhongjiajie replied:
done

songjianet reviewed (dismissed):
LGTM

caishunfeng commented on dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/ProcessDefinitionService.java:
```diff
@@ -67,6 +70,15 @@ Map<String, Object> createProcessDefinition(User loginUser,
+     * @param workflowCreateRequest the new workflow object will be created
+     * @return New ProcessDefinition object created just now
+     */
+    ProcessDefinition createProcessDefinitionV2(User loginUser, WorkflowCreateRequest workflowCreateRequest);
```
What's the different between v1 and v2? Maybe v2 it's not a good guide for developers if v1 can not be replaced with v2.

zhongjiajie replied:
v1 is for currently API(operate workflow and task and task relation object), and V2 is only operate workflow object. V2 is mainly for RESTful api, which we only change one single object instand of too many object

caishunfeng replied:
Will we use v2 uniformly in the future? including UI or support restful api or pyds?

zhongjiajie replied:
I hope we can do that, and in that moment, maybe it it the best time to migrate python api into separate repository

ruanwenjun commented on dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/WorkflowV2Controller.java:
```diff
@@ -0,0 +1,168 @@
+    @ResponseStatus(HttpStatus.CREATED)
+    @ApiException(CREATE_PROCESS_DEFINITION_ERROR)
+    @AccessLogAnnotation(ignoreRequestArgs = "loginUser")
+    public ResourceResponse createWorkflows(@ApiIgnore @RequestAttribute(value = Constants.SESSION_USER) User loginUser,
```
```suggestion
    public Result<ProcessDefinition> createWorkflows(@ApiIgnore @RequestAttribute(value = Constants.SESSION_USER) User loginUser,
```

ruanwenjun commented on dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/WorkflowV2Controller.java:
```diff
@@ -0,0 +1,168 @@
+    @ResponseStatus(HttpStatus.OK)
+    @ApiException(UPDATE_PROCESS_DEFINITION_ERROR)
+    @AccessLogAnnotation(ignoreRequestArgs = "loginUser")
+    public ResourceResponse updateWorkflows(@ApiIgnore @RequestAttribute(value = Constants.SESSION_USER) User loginUser,
```
```suggestion
    public Result<ProcessDefinition> updateWorkflows(@ApiIgnore @RequestAttribute(value = Constants.SESSION_USER) User loginUser,
```

ruanwenjun commented on dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/WorkflowV2Controller.java:
```diff
@@ -0,0 +1,168 @@
+    @ResponseStatus(HttpStatus.OK)
+    @ApiException(QUERY_PROCESS_DEFINITION_LIST)
+    @AccessLogAnnotation(ignoreRequestArgs = "loginUser")
+    public ResourceResponse getWorkflows(@ApiIgnore @RequestAttribute(value = Constants.SESSION_USER) User loginUser,
```
```suggestion
    public Result<ProcessDefinition> getWorkflows(@ApiIgnore @RequestAttribute(value = Constants.SESSION_USER) User loginUser,
```

zhongjiajie replied:
done

ruanwenjun commented on dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/WorkflowV2Controller.java:
```diff
@@ -0,0 +1,168 @@
+    @ResponseStatus(HttpStatus.OK)
+    @ApiException(QUERY_PROCESS_DEFINITION_LIST)
+    @AccessLogAnnotation(ignoreRequestArgs = "loginUser")
+    public PageResourceResponse filterWorkflows(@ApiIgnore @RequestAttribute(value = Constants.SESSION_USER) User loginUser,
```
```suggestion
    public Result<PageInfo<ProcessDefinition>> filterWorkflows(@ApiIgnore @RequestAttribute(value = Constants.SESSION_USER) User loginUser,
```

zhongjiajie replied:
done

ruanwenjun commented on dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/dto/ResourceResponse.java:
```diff
@@ -0,0 +1,39 @@
+ * workflow response
+ */
+@Data
+public class ResourceResponse<T> extends Result<T> {
```
It seems this pojo is same with Result.

zhongjiajie replied:
done

Tianqi-Dotes commented on dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/SchedulerServiceImpl.java:
```diff
@@ -270,6 +350,61 @@ public Map<String, Object> updateSchedule(User loginUser,
+     */
+    @Override
+    @Transactional
+    public Schedule getSchedules(User loginUser,
```
->getSchedule

zhongjiajie replied:
done

zhongjiajie committed:
Update dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/controller/ScheduleV2Controller.java

Co-authored-by: caishunfeng <caishunfeng2021@gmail.com>

zhongjiajie committed:
Fix according to suggestion

zhongjiajie commented:
we still have discuss in https://github.com/apache/dolphinscheduler/pull/11912#discussion_r973823553 and https://github.com/apache/dolphinscheduler/pull/11912#discussion_r973825097

zhongjiajie committed:
fix license issue and format code

zhongjiajie committed:
Merge remote-tracking branch 'upstream/dev' into f-openapi-wf-schedule

zhongjiajie committed:
change method name
from createProcessDefinitionV2 to createSingleProcessDefinition
from updateProcessDefinitionV2 to updateSingleProcessDefinition

caishunfeng reviewed (approved):
LGTM overall

caishunfeng commented:
Thanks @zhongjiajie
